R/rromeo_functions.R

```diff
@@ -67,18 +67,102 @@ rr_publisher = function(given_id) {
+rr_journal_issn = function(issn) {
 
   validate_issn(issn)
 
-  GET(rr_base_api(), query = list(issn = issn))
+  api_answer = GET(rr_base_api(), query = list(issn = issn))
+
+  parse_answer(api_answer, multiple = FALSE)
 }
 
+#' Journal data by title
+#'
+#' @param name A character string, containing the (possibly) partial name of the
+#' journal
+#' @param qtype A character string saying whether you are lookin for `exact`,
+#' `contains` or `starts with` matches
+#' @inheritParams parse_answer
+#'
+#' @export
+#'
+#' @examples
+#' rr_journal_name("Journal of Geology")
+#' rr_journal_name("Biogeography", multiple = FALSE, qtype = "contains")
+#' rr_journal_name("Biogeography", multiple = TRUE, qtype = "contains")
+rr_journal_name = function(name, multiple = FALSE,
+                           qtype = c("exact", "contains", "starts with")) {
+
+  qtype = match.arg(qtype)
+
+  api_answer = GET(rr_base_api(), query = list(jtitle = name, qtype = qtype))
+
+  parse_answer(api_answer, multiple = multiple)
+}
+
+#' Parse API answer.
+#'
+#' @param multiple If multiple results match your query, should the function
+#' recursively fetch data for each of one of them (`multiple = TRUE`) or
+#' return a data.frame containing only titles and ISSN of all matches
+#' (`multiple = FALSE`)
+#'
+#' @keywords internal
+#'
+#' @import xml2
+parse_answer = function(api_answer, multiple = FALSE) {
+
+  if (http_error(api_answer)) {
+    stop("The API endpoint could not be reached. Please try again later.")
+  }
+
+  xml_source = content(api_answer, encoding = "ISO-8859-1")
+
+  hits = xml_text(xml_find_all(xml_source, "//numhits"))
+
+  if (hits == 0) {
+    stop("No journal matches your query terms. Please try another query.")
+  }
+  else if (hits == 1) {
+    romeocolour = xml_text(xml_find_all(xml_source, "//romeocolour"))
+    preprint = xml_text(xml_find_all(xml_source, "//prearchiving"))
+    postprint = xml_text(xml_find_all(xml_source, "//postarchiving"))
+    pdf = xml_text(xml_find_all(xml_source, "//pdfarchiving"))
+
+    # TODO: which characteristics should we return?
+    return(data.frame(preprint, postprint, pdf, romeocolour))
+  }
+  else {
+    warning(hits, " journals match your query terms.\n")
+
+    if (xml_text(xml_find_all(xml_source, "//outcome")) == "excessJournals") {
+      warning("Your request exceeded SHERPA/RoMEO API's cap of 50 results. ",
+              "You should try to split your request into smaller chunks.")
+    }
+
+    journals = xml_text(xml_find_all(xml_source, "//jtitle"))
+    issns = xml_text(xml_find_all(xml_source, "//issn"))
+
+    if (!multiple) {
+      warning("Select one journal from the provided list or enable multiple = TRUE")
+
+      return(data.frame("Title" = journals,
+                        "ISSN" = issns))
+    } else {
+      message("Recursively fetching data from each journal. This may take some time...")
+
+      # FIXME: if data is missing for one of the journals, it is silently dropped.
```

Comment:
When does this happen? I mean if the journal has been found in SHERPA/RoMEO, can its data be missing?

Reply:
Actually, your suggested changes probably fix this. Before your changes, you could have a `NULL` in the results list because SHERPA/RoMEO had no info on preprint/postprint and pdf policy and therefore no colour as well. `NULL` elements are dropped when the list is transformed to a dataframe.

You can see an example of this in the vignette.

But now, all results will at least have a title (and most of the time, a issn as well) so it should be fine! I'll test it later.